open_instruct/dataset_processor.py

```diff
@@ -108,29 +108,31 @@
         "{% endfor %}"
     ),
     "zephyr": (
-        "{% for message in messages %}\n"
-        "{% if message['role'] == 'user' %}\n"
-        "{{ '<|user|>\n' + message['content'] + eos_token }}\n"
-        "{% elif message['role'] == 'system' %}\n"
-        "{{ '<|system|>\n' + message['content'] + eos_token }}\n"
-        "{% elif message['role'] == 'assistant' %}\n"
-        "{{ '<|assistant|>\n'  + message['content'] + eos_token }}\n"
-        "{% endif %}\n"
-        "{% if loop.last and add_generation_prompt %}\n{{ '<|assistant|>' }}\n"
-        "{% endif %}\n"
+        "{% for message in messages %}"
+        "{% if message['role'] == 'user' %}"
+        "{{ '<|user|>\n' + message['content'] + eos_token + '\n' }}"
+        "{% elif message['role'] == 'system' %}"
+        "{{ '<|system|>\n' + message['content'] + eos_token + '\n' }}"
+        "{% elif message['role'] == 'assistant' %}"
+        "{{ '<|assistant|>\n'  + message['content'] + eos_token + '\n' }}"
+        "{% endif %}"
+        "{% if loop.last and add_generation_prompt %}"
+        "{{ '<|assistant|>\n' }}"
+        "{% endif %}"
```

Comment:
So the main changes here is "not including a `\n` at `{% endif %}`"?

Reply:
I moved the necessary `\n` to inside the double brackets. This change is mainly for reducing the confusion in the templates, as the `\n` outside blocks should be trimmed, according to this [codebase](https://github.com/huggingface/transformers/blob/98adf24883b007c2a7fb17bab1c01b1614673433/src/transformers/utils/chat_template_utils.py#L413). But this confuses me with the real `\n` that we want to keep.

For example, I changed `\n{{ '<|assistant|>' }}\n` to `{{ '<|assistant|>\n' }}`, as we want to keep the `\n` in the end as part of the generation prompt (the old template also did this, but it's just not very clear whether they are trimmed or not).

Reply:
Makes sense! Nice change!